library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/ChatC2SMessage.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2023 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.chat.api.types;
+
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.network.message.MessageSignature;
+import net.minecraft.network.message.MessageSignatureList;
+import net.minecraft.network.packet.c2s.play.ChatMessageC2SPacket;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.quiltmc.qsl.chat.api.QuiltMessageType;
+import org.quiltmc.qsl.chat.impl.InternalQuiltChatApiUtil;
+
+import java.time.Instant;
+import java.util.EnumSet;
+
+public class ChatC2SMessage extends AbstractChatMessage<ChatMessageC2SPacket> {
+	private final String message;
+	private final Instant timestamp;
+	private final long salt;
+	private final @Nullable MessageSignature signature;
+	private final MessageSignatureList.Acknowledgment messageAcknowledgments;
+
+	public ChatC2SMessage(PlayerEntity player, boolean isOnClientSide, ChatMessageC2SPacket packet) {
+		this(
+				player,
+				isOnClientSide,
+				packet.message(),
+				packet.timestamp(),
+				packet.salt(),
+				packet.signature(),
+				packet.messageAcknowledgments()
+		);
```

Comment:
hm, I don't like the two tabs here, but that's the thing: I don't know if one tab is a standard that is being used here or if it's just Ennui being Ennui here :P
yeah, check other QSL code first, this might just be ennui being ennui here

Reply:
`QuiltBuiltinResourcePackProfile` seems to use double tabs at least, so ill leave it

![image](https://user-images.githubusercontent.com/52360088/212959993-2305d02a-8635-440b-8c26-039cef492903.png)

Reply:
Ah, I see